releng/release-tool/release-tool.go

```diff
@@ -0,0 +1,1186 @@
+	promoteRC := flag.String("promote-release-candidate", "", "The tag of an rc release that will be promoted to an official release")
+
+	autoRelease := flag.Bool("auto-release", false, "Automatically perform branch cutting an releases based on time intervals")
+	autoReleaseCadance := flag.String("auto-release-cadance", "monthly", "set the auto release cadance to daily or monthly")
+	autoPromoteAfterDays := flag.Int("auto-promote-after-days", 7, "Set the set the time before autopromoting a release candidate")
+
+	flag.Parse()
+
+	if *org == "" {
+		log.Fatal("--org is a required argument")
+	} else if *repo == "" {
+		log.Fatal("--repo is a required argument")
+	} else if *gitUser == "" {
+		log.Fatal(" --git-user is required")
+	} else if *gitEmail == "" {
+		log.Fatal("--git-email is required")
+	} else if *githubTokenFile == "" {
+		log.Fatal("--github-token-file is a required argument")
+	}
+
+	tokenBytes, err := ioutil.ReadFile(*githubTokenFile)
+	if err != nil {
+		log.Fatal(fmt.Printf("ERROR accessing github token: %s ", err))
+	}
+	token := strings.TrimSpace(string(tokenBytes))
+
+	repoUrl := fmt.Sprintf("https://%s@github.com/%s/%s.git", token, *org, *repo)
+	infraUrl := fmt.Sprintf("https://%s@github.com/kubevirt/project-infra.git", token)
+	repoDir := fmt.Sprintf("%s/%s/https-%s", *cacheDir, *org, *repo)
+	infraDir := fmt.Sprintf("%s/%s/https-%s", *cacheDir, "kubevirt", "project-infra")
+
+	if *cleanCacheDir {
+		os.RemoveAll(repoDir)
+		os.RemoveAll(infraDir)
+	}
+
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+
+	client := github.NewClient(tc)
+
+	r := releaseData{
+		repoDir:          repoDir,
+		infraDir:         infraDir,
+		repoUrl:          repoUrl,
+		infraUrl:         infraUrl,
+		repo:             *repo,
+		org:              *org,
+		newBranch:        *newBranch,
+		tag:              *releaseTag,
+		promoteRC:        *promoteRC,
+		skipReleaseNotes: *skipReleaseNotes,
+
+		gitUser:  *gitUser,
+		gitEmail: *gitEmail,
+		gitToken: token,
+
+		githubClient:    client,
+		githubTokenPath: *githubTokenFile,
+		dryRun:          *dryRun,
+		force:           *force,
+
+		blockerListCache: make(map[string]*blockerListCacheEntry),
+
+		now: time.Now(),
+	}
+
+	_, err = gitCommand("config", "--global", "user.name", r.gitUser)
+	if err != nil {
+		log.Fatal(err)
+	}
+	_, err = gitCommand("config", "--global", "user.email", r.gitEmail)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if *autoRelease {
+		r.autoDetectData(*autoReleaseCadance, *autoPromoteAfterDays)
+	}
+
+	// If this is a promotion, we need to set the tag to promote
+	if r.promoteRC != "" {
+		if r.tag != "" {
+			log.Fatal("--release-tag and --promote-rc can not be used together. --promote-rc detects the correct tag to make an official release out of")
+		}
+
+		re := regexp.MustCompile(`^v\d*\.\d*.\d*-rc.\d*$`)
```

Comment:
Maybe it's better to try and instantiate a `semver.Version` and let the semver library to handle the errors/compile.

Reply:
this regex is actually to enforce that we only promote an rc which is structured the way we expect it `<major>.<minor>.<patch>-rc.<rc version>`. It has to match this regardless if it is a valid semver or not.